aidial_adapter_openai/gpt4_multi_modal/transformation.py

```diff
@@ -0,0 +1,129 @@
+from typing import List, Optional, cast
+
+from pydantic import BaseModel, root_validator
+
+from aidial_adapter_openai.gpt4_multi_modal.download import (
+    ImageFail,
+    download_image,
+)
+from aidial_adapter_openai.gpt4_multi_modal.image_tokenizer import (
+    tokenize_image,
+)
+from aidial_adapter_openai.gpt4_multi_modal.messages import (
+    create_image_message,
+    create_text_message,
+)
+from aidial_adapter_openai.utils.image_data_url import ImageDataURL
+from aidial_adapter_openai.utils.log_config import logger
+from aidial_adapter_openai.utils.storage import FileStorage
+from aidial_adapter_openai.utils.tokens import Tokenizer
+
+
+class MessageTransformResult(BaseModel):
+    text_tokens: int = 0
+    image_tokens: int = 0
+    errors: Optional[List[ImageFail]] = None
+    message: dict
+
+    @property
+    def total_tokens(self) -> int:
+        return self.text_tokens + self.image_tokens
+
+
+class TransformMessagesResult(BaseModel):
+    error_message: Optional[str] = None
+    transformations: Optional[List[MessageTransformResult]] = None
+
+    @root_validator(pre=True)
+    def validate_info(cls, values):
+        if (
+            values.get("error_message") is None
+            and values.get("transformations") is None
+        ):
+            raise ValueError(
+                "Either 'error_message' or 'transformations' must be provided."
+            )
+        return values
+
+
+class TruncateTransformedMessagesResult(BaseModel):
+    messages: List[dict]
+    discarded_messages: List[int]
+    overall_token: int
+
+
+async def transform_message(
+    file_storage: Optional[FileStorage], message: dict, tokenizer: Tokenizer
+) -> MessageTransformResult:
+    content = message.get("content", "")
+    custom_content = message.get("custom_content", {})
+    attachments = custom_content.get("attachments", [])
+
+    message = {k: v for k, v in message.items() if k != "custom_content"}
+
+    if len(attachments) == 0:
+        return MessageTransformResult(
+            message=message,
+            text_tokens=tokenizer.calculate_tokens_per_message(message),
+        )
```

Comment:
This shortcut seems redundant. Nothing really would change performace-wise, semantic-wise if you remove it. The best code is one that is never written.

Reply:
Fixed, removed